utils/utf8.cc

```diff
@@ -87,8 +95,9 @@ static inline std::optional<size_t> validate_naive(const uint8_t *data, size_t l
                      (byte1 == 0xED && byte2 <= 0x9F) ||
                      // EE..EF, 80..BF, 80..BF
                      (byte1 >= 0xEE && byte1 <= 0xEF))) {
-                bytes = 3;
-            } else if (len >= 4) {
+                return codepoint_status{.bytes_validated = codepoint_len};
```

Comment:
I have to admit, I don't understand those checks. For example `(byte1 == 0xE0 && byte2 >= 0xA0)` why is `byte2 >= 0xA0` needed? I would understand `byte2 >= 0x80` but have no idea what `0xA0` is.
Same with the comparison of byte1 against `0xE1`, `0xED`, and `0xEE`. I thought that meaningful values of byte1 are `0x00 - 0x7F`, `0xC0-0xDF`, `0xE0-0xEF`, and `0xF0-0xF7`?

Reply:
I don't understand this either. For 3 bytes, byte0 should have binary 1110..., in other words between 0xE0 and 0xEF. What's all these special cases for 0xED and 0xEE?
All this code seems to me like maybe we're overthinking things... IIRC, in UTF-8, the first byte determines the length (that's the codepoint_len thing), but then - the rest of the bytes just need to begin with bits 10, i.e., between 0x80 and 0xBF.  That's it. Why do we have any other case?

Reply:
First, these comments don't concern the refactoring. I did not change the logic, just the interfaces.

Second, the code is very careful and correct. While the first byte determines the length, the following bytes have to follow a distinct pattern, and if they don't, the sequence is invalid. They don't just have to fit the range 0x80-0xbf.

I recommend you study the code, not the changes, if you want to understand the validation logic. My series does not change validation in any way (at least not intentionally), it just makes it work for scattered buffers.